### PR TITLE
Homework1

### PR DESCRIPTION
**Pull request by spin-to-win**
hw #1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,9 +1,51 @@
...
+```
+
+Тесты
+```
+python -m pytest -v
+```
+
+Самооценка:
+Я сделал все пункты кроме 7 и 12 из чеклиста https://data.mail.ru/blog/topic/view/18519/
+Соответственно вернхяя оценка 30 баллов, без учета штрафов за косяки, замечения, ошибки ну и так далее - это отдаю на суд ревьюера
```
Comment (kindofman) [inline README.md]: Для засчитывания 13-ого пункта необходима была точечная оценка.

File: ml_project/notebooks/01-exploratory_data_analysis.ipynb [hunk 2]
```diff
@@ -0,0 +1,1997 @@
...
+    "\n",
+    "from sklearn.ensemble import RandomForestClassifier \n",
+    "from sklearn.metrics import roc_curve, auc \n",
+    "from sklearn.metrics import classification_report \n",
+    "from sklearn.metrics import confusion_matrix \n",
+    "from sklearn.model_selection import train_test_split \n",
+    "\n",
+    "np.random.seed(123) \n",
+    "\n",
+    "pd.options.mode.chained_assignment = None     # hide any pandas warnings"
```
Comment (kindofman) [inline ml_project/notebooks/01-exploratory_data_analysis.ipynb]: Не рекомендую так делать. Я бы постарался написать код так, чтобы SettingWithCopyWarning не пришлось отлавливать на верхнем уровне. Тем более, что и без этой строчки ноутбук запустился без ворнингов.

https://stackoverflow.com/questions/20625582/how-to-deal-with-settingwithcopywarning-in-pandas

File: README.md [hunk 3]
```diff
@@ -1,9 +1,51 @@
...
+```
+python -m venv .venv
+source .venv/bin/activate
+pip install -e  .
+```
 
-made: d.svirchkov
+Обучение (на дефолтном конфиге):
+```
+python heart_disease_source_code/train.py
```
Comment (kindofman) [inline README.md]: Выглядит, как будто после переименования папки забыл поправить heart_disease_source_code -> src.

Comment (sashastds) [inline README.md]: это видимо неактуально уже?
не нашёл такой директории

File: ml_project/data/output/example_predicts.csv [hunk 4]
```diff
@@ -0,0 +1,304 @@
+preds
```
Comment (sashastds) [inline ml_project/data/output/example_predicts.csv:1]: некритично, и в целом тут нет айдишников, но я бы на всякий случай писал рядом индексы исходного фрейма, как хоть какой-то идентификатор

File: ml_project/src/data/make_dataset.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
```
Comment (sashastds) [inline ml_project/src/data/make_dataset.py:1]: опять же некритично, но много пустых строк в конце файлов

File: ml_project/src/features/build_features.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        df['thalassemia'][df['thalassemia'] == 0] = 'no'
+        df['thalassemia'][df['thalassemia'] == 1] = 'normal'
+        df['thalassemia'][df['thalassemia'] == 2] = 'fixed_defect'
+        df['thalassemia'][df['thalassemia'] == 3] = 'reversable_defect'
+
+        return df
+    
+    @staticmethod
+    def process_categorical_features(df: pd.DataFrame) -> pd.DataFrame:
+        """ Заполнить пропуски и задать категории
```
Comment (sashastds) [inline ml_project/src/features/build_features.py:114]: прикольное решение с одной стороны, но с другой, если это категории не было в трейнсете, то и дерево сплитовать по ней не научится, так что с этой точки зрения st_slope == 3 ничем же не отличается от потенциального st_slope == 7, например, для модели с точки зрения полезной информации? но вот в последнем случае оно сломается ведь?
может имело смысл как-то все неизвестные значения маппить в одно неинформативное?

File: ml_project/src/train.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+from omegaconf import OmegaConf
+import pandas as pd
+import pickle
+import logging
+from pprint import pprint
+
+from src.data import read_data, split_train_val_data
+from src.features.build_features import PreprocessRawData
+from src.features.select_features import select_features
+from src.models.save_and_load_models import serialize_model, load_model
```
Comment (sashastds) [inline ml_project/src/train.py:15]: могу ошибаться, но вроде порядок импортов не по PEP немного

File: ml_project/src/features/build_features.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import OneHotEncoder
+from sklearn.base import TransformerMixin
+
+USE_COLUMNS = [
+    'age','ca','chol','cp','exang','fbs','oldpeak',
+    'restecg','sex','slope','target','thal','thalach','trestbps'
+]
+
+class PreprocessRawData(TransformerMixin):
```
Comment (sashastds) [inline ml_project/src/features/build_features.py]: немного неканоничное название класса, хотя и в CamelCase, но поскольку глагол, а не например RawDataPreprocessor, чуть сбивает с толку

File: ml_project/tests/features/test_build_features.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def test_raise_if_bad_columns(train_data_sample):
+    train_data_broken = train_data_sample.rename(columns={'age': 'birthdate'})
+    with pytest.raises(Exception) as e_info:
+        PreprocessRawData.check_raw_data(train_data_broken)
+
+    train_data_broken = train_data_sample.drop(columns={'target'})
+    with pytest.raises(Exception) as e_info:
+        PreprocessRawData.check_raw_data(train_data_broken)
```
Comment (sashastds) [inline ml_project/tests/features/test_build_features.py]: тесты без ассертов это так и задумано?
условно если pytest не упал сам по себе == прошло?

**Review by sashastds — COMMENTED** [review]
С самооценкой согласен, у меня запустилось, обсчиталось.
Еще из мелочей - мне лично в ревью писали, что пиклы не стоит класть в гит, хотя они и мелкие.
И я не уверен, что понял, что именно считать отчётом, но EDA есть.
С гидрой выглядит так, что глубоко разбирался, круто

File: ml_project/setup.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        "dataclasses==0.6",
+        "pyyaml==5.4.1",
+        "marshmallow-dataclass==8.3.0",
+        "pandas==1.1.5",
+        "hydra-core==1.0.6",
+        "omegaconf==2.0.6",
+        "pytest==6.1.2",
+        "pytest-cov==2.10.1",
+        "tox==3.23.1",
+        "Faker==8.1.4",
```
Comment (kindofman) [inline ml_project/setup.py:22]: Не хватает библиотек для запуска ноутбука. Jupyter, matplotlib, seaborn.

**Review by kindofman — CHANGES_REQUESTED** [review]
Название файла train.py не соответствует назначению файла. Также выполняются предсказания. Чтобы название не вводило в заблуждение, нужно поменять название ключевого скрипта.

File: ml_project/setup.py (hunk 10 above)
Comment (spin-to-win) [reply ml_project/setup.py:22]: Так и задумано, чтобы не раздувать setup. 

Переделал следующим образом: переконвертировал в html и удалил ipynb, чтобы не смущать никого

File: README.md (hunk 3 above)
Comment (spin-to-win) [reply README.md]: хм странно... не знаю... сейчас вроде исправил

File: ml_project/notebooks/01-exploratory_data_analysis.ipynb (hunk 2 above)
Comment (spin-to-win) [reply ml_project/notebooks/01-exploratory_data_analysis.ipynb]: Done

File: README.md (hunk 1 above)
Comment (spin-to-win) [reply README.md]: Done

Добавил точечную оценку в README.md

Comment (spin-to-win) [reply README.md]: """Название файла train.py не соответствует назначению файла. Также выполняются предсказания. Чтобы название не вводило в заблуждение, нужно поменять название ключевого скрипта."""

Done

Сделал отдельно функцию трейн, отдельно предикт

File: README.md (hunk 3 above)
Comment (spin-to-win) [reply README.md]: Да, заменил heart_disease_source_code -> src

File: ml_project/data/output/example_predicts.csv (hunk 4 above)
Comment (spin-to-win) [reply ml_project/data/output/example_predicts.csv:1]: Ack

File: ml_project/src/data/make_dataset.py (hunk 5 above)
Comment (spin-to-win) [reply ml_project/src/data/make_dataset.py:1]: Ack

File: ml_project/src/features/build_features.py (hunk 6 above)
Comment (spin-to-win) [reply ml_project/src/features/build_features.py:114]: Ack

""" но вот в последнем случае оно сломается ведь? """ 
Да так и задумано, чтобы st_slope == 3 отработал,  а == 7 упал. Потому что 3 еще можно было бы ожидать, просто примеров в трейне не было. А вот если где-то появится неожиданная категория, например, st_slope == 7 то на этом месте по итогу будет стоять np.NaN и обучение упадет. 

"""может имело смысл как-то все неизвестные значения маппить в одно неинформативное?"""
Я правильно понимаю ты имеешь ввиду заменять неизвестные значения (типа st_slope == 7) на "unknown_category" и рейзить ошибку если найдены такие категории?

File: ml_project/src/features/build_features.py (hunk 8 above)
Comment (spin-to-win) [reply ml_project/src/features/build_features.py]: Done
Переименовал на RawDataPreprocessor

File: ml_project/tests/features/test_build_features.py (hunk 9 above)
Comment (spin-to-win) [reply ml_project/tests/features/test_build_features.py]: Да, просто смоук-тест (запустили и если не пошел дым, значит все хорошо)

**Review by kindofman — APPROVED** [review]
Отличная работа!

31 балл.
